c7n/tags.py

```diff
@@ -913,6 +914,48 @@ def process_resource_set(self, client, resource_set, tag_keys):
             client.untag_resources, ResourceARNList=arns, TagKeys=tag_keys)
 
 
+class UniversalTagRename(Action):
+
+    schema = utils.type_schema(
+        'rename-tag',
+        old_keys={'type': 'array', 'items': {'type': 'string'}},
+        old_key={'type': 'string'},
+        new_key={'type': 'string'},
+    )
+
+    permissions = UniversalTag.permissions + UniversalUntag.permissions
+
+    def validate(self):
+        if 'old_keys' not in self.data and 'old_key' not in self.data:
+            raise PolicyValidationError(
+                f"{self.manager.ctx.policy.name}:{self.type} 'old_keys' or 'old_key' required")
+
+    def process(self, resources):
+        old_keys = set(self.data.get('old_keys', ()))
+        if 'old_key' in self.data:
+            old_keys.add(self.data['old_key'])
+
+        # Collect by distinct tag value, and old_key value to minimize api calls
+        # via bulk api usage on add and remove.
+        old_key_resources = {}
+        values_resources = {}
+
+        for r in resources:
+            for t in r.get('Tags', ()):
+                if t['Key'] in old_keys:
+                    old_key_resources.setdefault(t['Key'], []).append(r)
+                    values_resources.setdefault(t['Value'], []).append(r)
```

Comment:
I think I see the flow here, we're making one bulk tag call for the new tag and one bulk untag call per old key. The piece I'm not sure about is the expected flow when a resource has multiple tags from `old_keys` defined.

It seems like if we're maximizing use of bulk tagging APIs, we're accepting that the "winning" value is effectively undefined/arbitrary when a resource has multiple old tags. Because in practice it's going to be "the latest value we saw when building `values_resources` that happens to be defined in a given resource".

Is that all accurate as you see it?

Reply:
Hmm thinking a little bit more, I suppose we _could_ bake in an expectation that the order of `old_keys` is significant. And then only add each resource to `values_resources` once (maybe we use the value from the first matching old key, but still remove all old keys?).

In either case wondering if it's worth documenting whatever expected behavior we go with to avoid some classes of questions/surprises later.

Reply:
also noting @thisisshi did an extant pr specific to asg for respecting source/policy defined order on precedence.

Reply:
thinking on this a bit more I think your suggestion for obeying precedence in old_keys makes sense, re populating value_resources only once per resource.

Reply:
Sounds good to me. Wondering if we also need an option to opt out of the untag pass (though I could see that being a separate change, don't want to keep layering complexity onto this one)